ic-canister-client/src/pocket_ic.rs

```diff
@@ -0,0 +1,160 @@
+    /// 1) `client` - The PocketIC client.
+    /// 2) `canister` - The canister principal.
+    /// 3) `caller` - The caller principal.
+    pub async fn with_client<F, R>(&self, f: F) -> R
+    where
+        F: Send + FnOnce(Arc<PocketIc>, Principal, Principal) -> R + 'static,
+        R: Send + 'static,
+    {
+        let client = self.client.clone();
+        let cansiter = self.canister;
+        let caller = self.caller;
+
+        tokio::task::spawn_blocking(move || f(client, cansiter, caller))
+            .await
+            .unwrap()
+    }
+
+    /// Performs update call with the given arguments.
+    pub async fn update<T, R>(&self, method: &str, args: T) -> CanisterClientResult<R>
+    where
+        T: ArgumentEncoder + Send + Sync,
+        R: for<'de> Deserialize<'de> + CandidType,
+    {
+        let args = candid::encode_args(args)?;
+        let method = String::from(method);
+
+        let call_result = self
+            .with_client(move |client, canister, caller| {
+                client.update_call(canister, caller, &method, args)
+            })
+            .await?;
+
+        let reply = match call_result {
+            WasmResult::Reply(reply) => reply,
+            WasmResult::Reject(e) => {
+                return Err(CanisterClientError::CanisterError((
+                    RejectionCode::CanisterError,
+                    e,
+                )));
+            }
+        };
+
+        let decoded = Decode!(&reply, R)?;
+        Ok(decoded)
+    }
+
+    /// Performs query call with the given arguments.
+    pub async fn query<T, R>(&self, method: &str, args: T) -> CanisterClientResult<R>
+    where
+        T: ArgumentEncoder + Send + Sync,
+        R: for<'de> Deserialize<'de> + CandidType,
+    {
+        let args = candid::encode_args(args)?;
+        let method = String::from(method);
+
+        let call_result = self
+            .with_client(move |env, canister, caller| {
+                env.query_call(canister, caller, &method, args)
+            })
+            .await?;
+
+        let reply = match call_result {
+            WasmResult::Reply(reply) => reply,
+            WasmResult::Reject(e) => {
+                return Err(CanisterClientError::CanisterError((
+                    RejectionCode::CanisterError,
+                    e,
+                )));
+            }
+        };
+
+        let decoded = Decode!(&reply, R)?;
+        Ok(decoded)
+    }
+}
+
+#[async_trait::async_trait]
+impl CanisterClient for PocketIcClient {
+    async fn update<T, R>(&self, method: &str, args: T) -> CanisterClientResult<R>
+    where
+        T: ArgumentEncoder + Send + Sync,
+        R: for<'de> Deserialize<'de> + CandidType,
+    {
+        PocketIcClient::update(self, method, args).await
+    }
+
+    async fn query<T, R>(&self, method: &str, args: T) -> CanisterClientResult<R>
+    where
+        T: ArgumentEncoder + Send + Sync,
+        R: for<'de> Deserialize<'de> + CandidType,
```

Comment:
Replace for<'de> Deserialize<'de> with DeserializeOwned

Reply:
Done